packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -172,111 +159,113 @@ export default function parse(source: string, state: State): TemplateParseResult
         }
     }
 
-    traverseHTML(templateRoot, {
```

Comment:
The main motivation of this PR is to convert the `traverseHTML` call into `parseElement`, `parseText` and `parseComment`. The recursion is handled by `parseChildren`.

By using the recursive approach, the `parent` and `stack` used to track the current AST node are not needed anymore.